X509_print_ex(): Do not abort printing when notBefore/notAfter times are invalid

DDvO opened the pull request:
When doing #13711 I found that the `-text` output of the `x509` app can stop prematurely in case the printed cert does not have notBefore or notAfter attributes, such as here:
```
Certificate:
    Data:
        Version: 1 (0x0)
        Serial Number:
            74:5f:dd:21:c4:74:e1:d4:6f:7f:27:76:22:a5:66:5c:29:e9:a9:77
        Signature Algorithm: sha256WithRSAEncryption
        Issuer: CN = t
        Validity
            Not Before: Bad time value-----BEGIN CERTIFICATE-----
MIICgTCCAWsCFHRf3SHEdOHUb38ndiKlZlwp6al3MAsGCSqGSIb3DQEBCzAMMQow
[...]
w1OUl8HhQdEOeiLPZg5IgujC1Jyn
-----END CERTIFICATE-----
```

This PR fixes the printing by making the output of `ASN1_TIME_print()` more consistent: 
it now returns failure only in case of an I/O error, 
also in case the input format is not valid, where anyway a suitable text output is provided: `Bad time value`.

kroeckx commented:
So the reason for changing it is really that there are 2 errors, where 1 is with the structure itself, the other an I/O error, and the caller can't tell the difference. You could argue that printing "Bad time value" is a successful call, which is what this PR does, but I have mixed feelings about it. You can also say that the caller should have called ASN1_TIME_check() first, and then deal with the error itself, for instance showing the string that it was unable to parse.

t8m commented:
@DDvO please do not edit comments to add a reply. It was seriously confusing to me to see who wrote what in the previous comment.

Could the problem be solved in X509_print_ex without changing the semantics of the ASN1_TIME_print()? Basically call the ASN1_TIME_check() first and if it fails then just call ASN1_TIME_print() anyway and ignore the error return?

Otherwise we might have to call an OTC vote for this as this is a hidden API break.

DDvO commented:
> @DDvO please do not edit comments to add a reply. It was seriously confusing to me to see who wrote what in the previous comment.

Oh, by mistake I had hit the "Edit" menu entry rather than "Quote reply" - sorry for that.
I've added a respective note to the above comment.

> Could the problem be solved in X509_print_ex without changing the semantics of the ASN1_TIME_print()? Basically call the ASN1_TIME_check() first and if it fails then just call ASN1_TIME_print() anyway and ignore the error return?

Something like this is possible, but I consider it pretty ugly since basically each call of `ASN1_TIME_print()` and the like needs to do such a check.
I've tentatively done so for the uses in `X509_print_ex()`.

> Otherwise we might have to call an OTC vote for this as this is a hidden API break.

In order to avoid this, I've reverted the change of return value for `ASN1_TIME_print()`,
adding a comment on its inconsistent error return behavior and 
adding a hint for a workaround using `ASN1_TIME_check()` etc. in the documentation.

t8m commented on crypto/x509/t_x509.c:
```diff
@@ -45,6 +45,12 @@ int X509_print(BIO *bp, X509 *x)
 }
 
+/* Prints some string in any case and return error only on I/O error */
+static int ASN1_TIME_print_also_invalid(BIO *bp, const ASN1_TIME *tm)
```
Please use something like `asn1_time_print_ignore_invalid`.

DDvO replied:
It turns out that using `ASN1_TIME_check()` is not the right solution.
Its error checking is different to `asn1_time_to_tm()` used by  `ASN1_TIME_print()`.

Changed the solution to introducing a cryptolib-internal function `ASN1_TIME_print_ex()`
with the desired behavior while retaining the original semantics for `ASN1_TIME_print()`.

If this is fine by you I'll extend its use to further internal calls of `ASN1_TIME_print()`.

t8m reviewed (commented):
Just one nit - I think the function name is confusing and it also should be in lower case because it is static function.

t8m commented on crypto/asn1/a_time.c:
```diff
@@ -468,6 +469,11 @@ static const char _asn1_mon[12][4] = {
+    return ASN1_TIME_print_ex(bp, tm, 0);
+}
+
+int ASN1_TIME_print_ex(BIO *bp, const ASN1_TIME *tm, int also_invalid)
```
What about changing semantics of the return value rather than adding parameter? Return -1 in case of parse failure, 0 on BIO write error.

Also the name should be `asn1_time_print_ex` as the function is internal.

DDvO replied:
Good thought that changing the return value is more direct. 
Done; also made the function name lower-case.

DDvO commented on crypto/asn1/a_time.c:
```diff
@@ -476,7 +482,7 @@ int ASN1_TIME_print(BIO *bp, const ASN1_TIME *tm)
     if (!asn1_time_to_tm(&stm, tm)) {
         /* asn1_time_to_tm will check the time type */
-        goto err;
+        return BIO_write(bp, "Bad time value", 14) || also_invalid;
```
Ah, right `&&` would have been correct.
Yet as you suggested below the parameter is gone now.

t8m reviewed (approved):
I've adjusted the title of this PR to make it clear what is the fix about now.

Good work!

t8m commented:
Please adjust the commit message of the second commit when merging to reflect the actual change after the review.

DDvO commented:
> Please adjust the commit message of the second commit when merging to reflect the actual change after the review.

I've just done this such that I won't miss it when merging.
Thanks also for the other improvement hints.

openssl-machine commented:
24 hours has passed since 'approval: done' was set, but as this PR has been updated in that time the label 'approval: ready to merge' is not being automatically set.  Please review the updates and set the label manually.

DDvO commented:
Merged - thanks @t8m

FdaSilvaYY commented on doc/man3/ASN1_TIME_set.pod:
```diff
@@ -199,9 +202,9 @@ ASN1_TIME_normalize() returns 1 on success, and 0 on error.
-or invalid time format).
+ASN1_TIME_print(), ASN1_UTCTIME_print() and ASN1_GENERALIZEDTIME_print()
+return 1 if the time is successfully printed out and
+0 if an I/O error occurred an error occurred (I/O error or invalid time format).
```
kind of copy/paste mistake on this line.
Sorry for the delayed reaction.

DDvO replied:
Thanks @FdaSilvaYY for pointing this out.
I've just simply added a small commit to #13735 for fixing the glitch.
